File: backend/app/models/sheets.py
from __future__ import annotations

from pydantic import BaseModel, Field


class GlideParams(BaseModel):
    start_equity: float = 0.6
    trough_equity: float = 0.4
    end_equity: float = 0.6
    recovery_years: int = 10


class WithdrawalParams(BaseModel):
    initial_rate: float = 0.04
    upper_pct: float = 0.20
    lower_pct: float = 0.20
    adjustment_pct: float = 0.10
    prosperity_years: int = 15


class CashFlow(BaseModel):
    start_age: int
    monthly_real: float
    cola: bool = True


class OneOff(BaseModel):
    age: int
    amount_real: float


class Assumptions(BaseModel):
    """Parsed contents of the `Assumptions` tab in the user's sheet."""

    current_age: int | None = None
    retirement_age: int | None = None
    end_age: int | None = None
    annual_spend_real: float | None = None
    annual_contributions: float | None = None
    swr: float | None = None
    equity_mean: float | None = None
    equity_sd: float | None = None
    bond_mean: float | None = None
    bond_sd: float | None = None
    correlation: float | None = None
    glide_kind: str | None = None
    glide_params: GlideParams | None = None
    withdrawal_kind: str | None = None
    withdrawal_params: WithdrawalParams | None = None
    social_security: CashFlow | None = None
    pensions: list[CashFlow] = Field(default_factory=list)
    one_offs: list[OneOff] = Field(default_factory=list)


class SheetsStatus(BaseModel):
    authorized: bool
    sheet_id: str | None
    last_write_at: str | None = None
    error: str | None = None


class TargetRow(BaseModel):
    age: int
    target_net_worth: float


class PushResult(BaseModel):
    written_rows: int
    range: str

File: frontend/src/lib/api/sheets.ts
import { api } from "./client";

export interface GlideParams {
  start_equity: number;
  trough_equity: number;
  end_equity: number;
  recovery_years: number;
}

export interface WithdrawalParams {
  initial_rate: number;
  upper_pct: number;
  lower_pct: number;
  adjustment_pct: number;
  prosperity_years: number;
}

export interface CashFlow {
  start_age: number;
  monthly_real: number;
  cola: boolean;
}

export interface OneOff {
  age: number;
  amount_real: number;
}

export interface Assumptions {
  current_age: number | null;
  retirement_age: number | null;
  end_age: number | null;
  annual_spend_real: number | null;
  annual_contributions: number | null;
  swr: number | null;
  equity_mean: number | null;
  equity_sd: number | null;
  bond_mean: number | null;
  bond_sd: number | null;
  correlation: number | null;
  glide_kind: string | null;
  glide_params: GlideParams | null;
  withdrawal_kind: string | null;
  withdrawal_params: WithdrawalParams | null;
  social_security: CashFlow | null;
  pensions: CashFlow[];
  one_offs: OneOff[];
}

export interface SheetsStatus {
  authorized: boolean;
  sheet_id: string | null;
  last_write_at: string | null;
  error: string | null;
}

export interface TargetRow {
  age: number;
  target_net_worth: number;
}

export interface PushResult {
  written_rows: number;
  range: string;
}

export const sheetsApi = {
  status: () => api<SheetsStatus>("/api/sheets/status"),
  assumptions: () => api<Assumptions>("/api/sheets/assumptions"),
  targets: () => api<TargetRow[]>("/api/sheets/targets"),
  readTab: (name: string) => api<unknown[][]>(`/api/sheets/tab/${encodeURIComponent(name)}`),
  pushAccounts: () => api<PushResult>("/api/sheets/push/accounts", { method: "POST" }),
  pushHoldings: () => api<PushResult>("/api/sheets/push/holdings", { method: "POST" }),
  pushProjection: (result: unknown) =>
    api<PushResult>("/api/sheets/push/projection", {
      method: "POST",
      body: JSON.stringify(result),
    }),
};
